{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "improving-flour",
   "metadata": {},
   "source": [
    "## Fixes:\n",
    "- sort in absolute value\n",
    "- Put it in a module / pip / conda\n",
    "- Add differential expression feature\n",
    "- Calculate everything in matrices rather than for loops\n",
    "- When you color cells, what should the max and min expression be? Based on the selected cells or based on all of the cells. Currently it seems to remap based on only the selected cells but this is maybe misleading."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "caroline-break",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "lesser-haven",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext py5\n",
    "%gui osx\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "younger-invention",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print('DOWNLOADING AND EXTRACTING EXAMPLE DATA')\n",
    "#! mkdir -p ../data\n",
    "#! wget https://storage.googleapis.com/sabeti-public/dkotliar/scnavigator/pbmc3k/data/pbmc3k_umap.tsv -O ../data/pbmc3k_umap.tsv\n",
    "#! wget https://storage.googleapis.com/sabeti-public/dkotliar/scnavigator/pbmc3k/data/pbmc3k_expression_filtered_normalized.tsv.gz -O ../data/pbmc3k_expression_filtered_normalized.tsv.gz\n",
    "#! gzip -df ../data/pbmc3k_expression_filtered_normalized.tsv.gz\n",
    "#! ls ../data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "foreign-princeton",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOADING UMAP DATA...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UMAP_1</th>\n",
       "      <th>UMAP_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATACAACCAC-1</th>\n",
       "      <td>3.638991</td>\n",
       "      <td>3.122167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGAGCTAC-1</th>\n",
       "      <td>-0.490064</td>\n",
       "      <td>10.553852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGATCAGC-1</th>\n",
       "      <td>0.391078</td>\n",
       "      <td>4.035561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGCTTCCG-1</th>\n",
       "      <td>-6.747729</td>\n",
       "      <td>-0.600776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGTATGCG-1</th>\n",
       "      <td>1.887045</td>\n",
       "      <td>-1.526383</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    UMAP_1     UMAP_2\n",
       "index                                \n",
       "AAACATACAACCAC-1  3.638991   3.122167\n",
       "AAACATTGAGCTAC-1 -0.490064  10.553852\n",
       "AAACATTGATCAGC-1  0.391078   4.035561\n",
       "AAACCGTGCTTCCG-1 -6.747729  -0.600776\n",
       "AAACCGTGTATGCG-1  1.887045  -1.526383"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"LOADING UMAP DATA...\")\n",
    "\n",
    "_umap = pd.read_csv('../data/pbmc3k_umap.tsv', sep='\\t')\n",
    "_umap.index = _umap['index']\n",
    "_umap = _umap.iloc[:, 1:]\n",
    "_umap.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "canadian-italy",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOADING GENE EXPRESSION DATA...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TNFRSF4</th>\n",
       "      <th>CPSF3L</th>\n",
       "      <th>ATAD3C</th>\n",
       "      <th>C1orf86</th>\n",
       "      <th>RER1</th>\n",
       "      <th>TNFRSF25</th>\n",
       "      <th>TNFRSF9</th>\n",
       "      <th>CTNNBIP1</th>\n",
       "      <th>SRM</th>\n",
       "      <th>UBIAD1</th>\n",
       "      <th>...</th>\n",
       "      <th>DSCR3</th>\n",
       "      <th>BRWD1</th>\n",
       "      <th>BACE2</th>\n",
       "      <th>SIK1</th>\n",
       "      <th>C21orf33</th>\n",
       "      <th>ICOSLG</th>\n",
       "      <th>SUMO3</th>\n",
       "      <th>SLC19A1</th>\n",
       "      <th>S100B</th>\n",
       "      <th>PRMT2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATACAACCAC-1</th>\n",
       "      <td>-0.171470</td>\n",
       "      <td>-0.280812</td>\n",
       "      <td>-0.046677</td>\n",
       "      <td>-0.475169</td>\n",
       "      <td>-0.544024</td>\n",
       "      <td>4.928496</td>\n",
       "      <td>-0.038028</td>\n",
       "      <td>-0.280573</td>\n",
       "      <td>-0.341788</td>\n",
       "      <td>-0.195361</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.226570</td>\n",
       "      <td>-0.236269</td>\n",
       "      <td>-0.102943</td>\n",
       "      <td>-0.222116</td>\n",
       "      <td>-0.312401</td>\n",
       "      <td>-0.121678</td>\n",
       "      <td>-0.521229</td>\n",
       "      <td>-0.098269</td>\n",
       "      <td>-0.209095</td>\n",
       "      <td>-0.531203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGAGCTAC-1</th>\n",
       "      <td>-0.214582</td>\n",
       "      <td>-0.372653</td>\n",
       "      <td>-0.054804</td>\n",
       "      <td>-0.683391</td>\n",
       "      <td>0.633951</td>\n",
       "      <td>-0.334837</td>\n",
       "      <td>-0.045589</td>\n",
       "      <td>-0.498264</td>\n",
       "      <td>-0.541914</td>\n",
       "      <td>-0.209017</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.317531</td>\n",
       "      <td>2.568868</td>\n",
       "      <td>0.007155</td>\n",
       "      <td>-0.445372</td>\n",
       "      <td>1.629285</td>\n",
       "      <td>-0.058662</td>\n",
       "      <td>-0.857163</td>\n",
       "      <td>-0.266844</td>\n",
       "      <td>-0.313146</td>\n",
       "      <td>-0.596654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGATCAGC-1</th>\n",
       "      <td>-0.376888</td>\n",
       "      <td>-0.295085</td>\n",
       "      <td>-0.057527</td>\n",
       "      <td>-0.520972</td>\n",
       "      <td>1.332648</td>\n",
       "      <td>-0.309362</td>\n",
       "      <td>-0.103108</td>\n",
       "      <td>-0.272526</td>\n",
       "      <td>-0.500799</td>\n",
       "      <td>-0.220228</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.302938</td>\n",
       "      <td>-0.239801</td>\n",
       "      <td>-0.071774</td>\n",
       "      <td>-0.297857</td>\n",
       "      <td>-0.410920</td>\n",
       "      <td>-0.070431</td>\n",
       "      <td>-0.590720</td>\n",
       "      <td>-0.158656</td>\n",
       "      <td>-0.170876</td>\n",
       "      <td>1.379000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGCTTCCG-1</th>\n",
       "      <td>-0.285241</td>\n",
       "      <td>-0.281735</td>\n",
       "      <td>-0.052227</td>\n",
       "      <td>-0.484929</td>\n",
       "      <td>1.572680</td>\n",
       "      <td>-0.271825</td>\n",
       "      <td>-0.074552</td>\n",
       "      <td>-0.258875</td>\n",
       "      <td>-0.416752</td>\n",
       "      <td>-0.208471</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.262978</td>\n",
       "      <td>-0.231807</td>\n",
       "      <td>-0.093818</td>\n",
       "      <td>-0.247770</td>\n",
       "      <td>2.552079</td>\n",
       "      <td>-0.097402</td>\n",
       "      <td>1.631684</td>\n",
       "      <td>-0.119462</td>\n",
       "      <td>-0.179120</td>\n",
       "      <td>-0.505669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGTATGCG-1</th>\n",
       "      <td>-0.256484</td>\n",
       "      <td>-0.220394</td>\n",
       "      <td>-0.046800</td>\n",
       "      <td>-0.345859</td>\n",
       "      <td>-0.333409</td>\n",
       "      <td>-0.208122</td>\n",
       "      <td>-0.069514</td>\n",
       "      <td>5.806440</td>\n",
       "      <td>-0.283112</td>\n",
       "      <td>-0.199355</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.202237</td>\n",
       "      <td>-0.176765</td>\n",
       "      <td>-0.167350</td>\n",
       "      <td>-0.098665</td>\n",
       "      <td>-0.275836</td>\n",
       "      <td>-0.139482</td>\n",
       "      <td>-0.310095</td>\n",
       "      <td>-0.006877</td>\n",
       "      <td>-0.109614</td>\n",
       "      <td>-0.461946</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1838 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   TNFRSF4    CPSF3L    ATAD3C   C1orf86      RER1  TNFRSF25  \\\n",
       "index                                                                          \n",
       "AAACATACAACCAC-1 -0.171470 -0.280812 -0.046677 -0.475169 -0.544024  4.928496   \n",
       "AAACATTGAGCTAC-1 -0.214582 -0.372653 -0.054804 -0.683391  0.633951 -0.334837   \n",
       "AAACATTGATCAGC-1 -0.376888 -0.295085 -0.057527 -0.520972  1.332648 -0.309362   \n",
       "AAACCGTGCTTCCG-1 -0.285241 -0.281735 -0.052227 -0.484929  1.572680 -0.271825   \n",
       "AAACCGTGTATGCG-1 -0.256484 -0.220394 -0.046800 -0.345859 -0.333409 -0.208122   \n",
       "\n",
       "                   TNFRSF9  CTNNBIP1       SRM    UBIAD1  ...     DSCR3  \\\n",
       "index                                                     ...             \n",
       "AAACATACAACCAC-1 -0.038028 -0.280573 -0.341788 -0.195361  ... -0.226570   \n",
       "AAACATTGAGCTAC-1 -0.045589 -0.498264 -0.541914 -0.209017  ... -0.317531   \n",
       "AAACATTGATCAGC-1 -0.103108 -0.272526 -0.500799 -0.220228  ... -0.302938   \n",
       "AAACCGTGCTTCCG-1 -0.074552 -0.258875 -0.416752 -0.208471  ... -0.262978   \n",
       "AAACCGTGTATGCG-1 -0.069514  5.806440 -0.283112 -0.199355  ... -0.202237   \n",
       "\n",
       "                     BRWD1     BACE2      SIK1  C21orf33    ICOSLG     SUMO3  \\\n",
       "index                                                                          \n",
       "AAACATACAACCAC-1 -0.236269 -0.102943 -0.222116 -0.312401 -0.121678 -0.521229   \n",
       "AAACATTGAGCTAC-1  2.568868  0.007155 -0.445372  1.629285 -0.058662 -0.857163   \n",
       "AAACATTGATCAGC-1 -0.239801 -0.071774 -0.297857 -0.410920 -0.070431 -0.590720   \n",
       "AAACCGTGCTTCCG-1 -0.231807 -0.093818 -0.247770  2.552079 -0.097402  1.631684   \n",
       "AAACCGTGTATGCG-1 -0.176765 -0.167350 -0.098665 -0.275836 -0.139482 -0.310095   \n",
       "\n",
       "                   SLC19A1     S100B     PRMT2  \n",
       "index                                           \n",
       "AAACATACAACCAC-1 -0.098269 -0.209095 -0.531203  \n",
       "AAACATTGAGCTAC-1 -0.266844 -0.313146 -0.596654  \n",
       "AAACATTGATCAGC-1 -0.158656 -0.170876  1.379000  \n",
       "AAACCGTGCTTCCG-1 -0.119462 -0.179120 -0.505669  \n",
       "AAACCGTGTATGCG-1 -0.006877 -0.109614 -0.461946  \n",
       "\n",
       "[5 rows x 1838 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"LOADING GENE EXPRESSION DATA...\")\n",
    "\n",
    "_expr = pd.read_csv('../data/pbmc3k_expression_filtered_normalized.tsv', sep='\\t')\n",
    "_expr.set_index('index', inplace=True)\n",
    "_expr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "conceptual-female",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../umap_explorer')\n",
    "from umap_explorer import UMAPexplorer\n",
    "\n",
    "test = UMAPexplorer(_umap, _expr)\n",
    "test.explore_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "characteristic-anaheim",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selected 618 cells\n",
      "Calculating correlations...\n",
      "Selected gene CCL5\n",
      "Min/max expression level for gene CCL5 -1.1439133882522583 3.2042019367218018\n",
      "Selected gene CST7\n",
      "Min/max expression level for gene CST7 -0.7983383536338806 3.1324713230133057\n",
      "Selected gene GZMA\n",
      "Min/max expression level for gene GZMA -0.9428274035453796 3.680380582809448\n"
     ]
    }
   ],
   "source": [
    "test.selected_gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "discrete-minority",
   "metadata": {},
   "outputs": [],
   "source": [
    "test.significant_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "recovered-delaware",
   "metadata": {},
   "outputs": [],
   "source": [
    "test.selected_cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sacred-visitor",
   "metadata": {},
   "outputs": [],
   "source": [
    "FCGR3A, CD79B, LINC00494    ... GPX1, LGALS2, S100A8"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
